cmd/ndm_daemonset/probe/samplingprobe.go

```diff
@@ -0,0 +1,212 @@
+
+	// if device is not a partition, check whether the zfs partitions are available on the disk
+	if blockDevice.DeviceAttributes.DeviceType != libudevwrapper.UDEV_PARTITION {
+		path, ok := getBlockDeviceZFSPartition(*blockDevice)
+		if ok {
+			devPath = path
+		} else {
+			hasZFS = false
+			klog.V(4).Infof("device: %s is not having any zfs partitions", blockDevice.DevPath)
+		}
+	}
+
+	if hasZFS {
+		samplingProbe := newSamplingProbe(devPath)
+
+		// check for ZFS file system
+		fstype := samplingProbe.BlkidIdentifier.GetOnDiskFileSystem()
+
+		if fstype == zfsFileSystemLabel {
+			blockDevice.DevUse.InUse = true
+
+			// the disk can either be in use by cstor or zfs local PV
+			ok, err := isBlockDeviceInUse(blockDevice.DevPath)
+
+			if err != nil {
+				klog.Errorf("error checking block device: %s: %v", blockDevice.DevPath, err)
+			}
+			if ok {
+				blockDevice.DevUse.UsedBy = blockdevice.ZFSLocalPV
+			} else {
+				blockDevice.DevUse.UsedBy = blockdevice.CStor
+			}
+			klog.V(4).Infof("device: %s Used by: %s filled by sampling probe", blockDevice.DevPath, blockDevice.DevUse.UsedBy)
+			return
+		}
+	}
+
+	// create a device identifier for reading the spdk super block from the disk
+	spdkIdentifier := &spdk.DeviceIdentifier{
+		DevPath: blockDevice.DevPath,
+	}
+
+	signature, err := spdkIdentifier.GetSPDKSuperBlockSignature()
+	if err != nil {
+		klog.Errorf("error reading spdk signature from device: %s, %v", blockDevice.DevPath, err)
+	}
+	if spdk.IsSPDKSignatureExist(signature) {
+		blockDevice.DevUse.InUse = true
+		blockDevice.DevUse.UsedBy = blockdevice.Mayastor
+		klog.V(4).Infof("device: %s Used by: %s filled by sampling probe", blockDevice.DevPath, blockDevice.DevUse.UsedBy)
+		return
+	}
+
+	// TODO jiva disk detection
+}
+
+// getBlockDeviceZFSPartition is used to get the zfs partition if it exist in a
+// given BD
+func getBlockDeviceZFSPartition(bd blockdevice.BlockDevice) (string, bool) {
+
+	// check for zfs partitions only if there are 2 partitions on the block device
+	if len(bd.DependentDevices.Partitions) != 2 {
+		return "", false
+	}
+
+	zfsDataPartitionNumber := "1"
+	zfsMetaPartitionNumber := "9"
+
+	// to handle cases of devices with nvme drives, the device name will be
+	// nvme0n1
+	if util.IsMatchRegex(".+[0-9]+$", bd.DevPath) {
+		zfsDataPartitionNumber = "p" + zfsDataPartitionNumber
+		zfsMetaPartitionNumber = "p" + zfsMetaPartitionNumber
+	}
+
+	dataPartition := bd.DevPath + zfsDataPartitionNumber
+	metaPartition := bd.DevPath + zfsMetaPartitionNumber
+
+	// check if device has the 2 partitions
+	if bd.DependentDevices.Partitions[0] == dataPartition &&
+		bd.DependentDevices.Partitions[1] == metaPartition {
+		return dataPartition, true
+	}
+	return "", false
+}
+
+// isBlockDeviceInUse tries to open the device exclusively to check if the device is
+// being held by some process. eg: If kernel zfs uses the disk, the open will fail
+func isBlockDeviceInUse(path string) (bool, error) {
```

Comment:
In both the cases, block device is in use. Can this function be renamed as isBlockDeviceInUseByKernel?

Reply:
Done.